### PR TITLE
Add basic engine

### PR DESCRIPTION
**Pull request by dombozzuto**
Added a basic game engine to handle ticks, and its pretty cool!
All agents now maintain a queue of actions and implement a `perform_actions` function that handles all of the actions in the queue.
- move action was moved from BaseAgent to it's own action class
- MoveAction is fully implemented, CollectAction created to show how new actions would be created
- check out the 5 new tests in test_game_engine.py to see how the engine works

Also added proper pydocs to almost everything; figured it would be helpful when the codebase is very volatile in its early days.

Cheers

### CODE REVIEW

File: server/src/agent/actions/impl/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from .collect import CollectAction
+from .move import MoveAction
+
+__all__ = [
+    'CollectAction',
+    'MoveAction'
+]
```
Comment (jrhackett) [inline server/src/agent/actions/impl/__init__.py]: End line!

File: server/src/map/map.py [hunk 2]
```diff
@@ -1,37 +1,68 @@
 from server.src.map import tiles
-from server.src.map.tiles import impl
+from server.src.map.resources import ResourceType
+from server.src.map.tiles.tile_factory import TileFactory
 
 
 class Map:
-
+    """
+    2x2 representation of the world in which the simulation will take place.
```
Comment (jrhackett) [inline server/src/map/map.py]: Nitpicking but maybe 2D is a better way to say 2x2 so there's no ambiguity on dimensions of the map and dimension of the world.

File: server/test/agent/test_agent_movement.py [hunk 3]
```diff
@@ -1,89 +1,71 @@
...
 import logging
+
 from server.src.agent.base_agent import BaseAgent
 from server.src.agent.agents import AgentType
 from server.src.agent.agent_factory import AgentFactory
 
 LOG = logging.getLogger(__name__)
 
 
+@pytest.mark.parametrize('agent_type', [
```
Comment (jrhackett) [inline server/test/agent/test_agent_movement.py]: Ooooo I like this.

**Comment by dombozzuto** [comment]
I think actions can eventually be expanded to have a duration element for multi-tick tasks, and after a tick of execution, the actions duration count gets decremented and is only removed from the agent's actions after it reaches 0

**Pull request merged**